### ISSUE TITLE
Rate iteration close to 0% bug

### ISSUE THREAD

**Issue by achan001** [thomasokken/plus42desktop#4]
Regarding **TVM I%YR** rate search ...

A rare bug, when initial guess, or one of the iteration, is *very* close to 0%
Catastrophic cancellations could mess up derivatives, see [here](https://www.hpmuseum.org/forum/thread-18359-post-161549.html#pid161549)

Instead of special cased to test _exactly_ i=0,  we widen it:
```
        local eps, f = i*i
        if 1+eps==1 then
            local a, b = (pv+fv)/n, pv-fv
            local fp, fpp = (a+b)*0.5, (n*n-1)*a/6
            f, fp = (a+pmt)+fp*i, fp+fpp*i
            eps = -f / fp
```

Above simply use an approximation, for tiny i

f(i) ≈ f(0) + f'(0)*i
f'(i) ≈ f'(0) + f''(0)*i

**Comment by achan001** [comment]
f(x) = f(0) + f'(0) * x + f''(0) * (x^2/2!) + f'''(0) * (x^3/3!) + ...

Let a=(pv+fv)/n, b=(pv-fv)
```
XCAS> g := n*x / ((1+x)^n-1)
XCAS> pv := (a*n+b)/2

XCAS> f := a*g + pv*x + pmt
XCAS> [f3,f2,f1,f0] := factor(e2r(series(f, x=0,3,polynom)))
XCAS> factor([f0, f1, f2*=2,f3*=6])

[a+pmt, (b+a)/2, (-1+n)*(1+n)*a/6, -(-1+n)*(1+n)*a/4]
```

**Comment by achan001** [comment]
To reduce chance of false negative, I widened fuzz factor

> return i-eps/2, (i == i-eps*0.001)

This may be especially needed for Plus42 Decimal version.
expm1 / log1p are not as accurate as true Decimal implementation.

Accuracy depends on size of condition number.
see [Decimal Transcendentals via Binary](http://www.acsel-lab.com/arithmetic/arith19/papers/ARITH19_Harrison.pdf)

**Comment by thomasokken** [comment]
I would like to fix this issue in the next release... Would you recommend using the code from the first post in the MoHPC Forum thread?

**Comment by achan001** [comment]
Patch for function **do_i_pct_yr()**, widen special case i=0, to tiny i  

f(i) ≈ f(0) + f'(0)*i
f'(i) ≈ f'(0) + f''(0)*i

```
<            if (i == 0) {
<                phloat a = (pv + fv) / n;
<                phloat b = pv - fv;
<                f = a + pmt;
<                phloat fp = (a + b) / 2;
<                i = eps = -f / fp;

>           if (1+i*i == 1) {
>                phloat a = (pv + fv) / n;
>                phloat b = pv - fv;
>                phloat fp = (a+b)*0.5, fpp = (n*n-1)*a/6;
>                f = a + pmt + fp*i;
>                fp += fpp*i;
>                i += (eps = -f / fp);
```

And, for bigger fuzz factor, if f(i) alone cannot decide convergence.

```
<                if (i == i - eps / 100) {
<                    i -= eps;
<                    break;

>                if (i == i - eps * 0.001) {
>                    i -= eps * 0.5;
>                    break;
```

**Comment by thomasokken** [comment]
I uploaded test builds for Windows to https://thomasokken.com/plus42/download/test/

**Comment by achan001** [comment]
Thank you!

**Comment by thomasokken** [comment]
My pleasure.
Thank you for your help!

**Comment by achan001** [comment]
I discovered a flaw with estimate (f, fp) close to 0%, using f(0), f(0)', f(0)''.
This in effect assumed curve is quadratic.

However, extremely small rate is almost always matched with big n, with huge compounding effect.

I tried to "fix it" with extra term, f = (a+pmt) + fp * i **+ fpp * i^2/2**.
It is better, but still not that good (compare to main branch)

Instead of (1+i*i == 1), I think it is better to use **(1+i == 1)**
In other words, branch only activate when main branch would fail.

**Comment by achan001** [comment]
I added the **f''(0)/2*i^2** term as well, for better accuracy.

Technically, we have direct solution for (f(0) + f'(0)*i + f''(0)*i^2/2) = 0
But algorithm assumed Newton's method for convergence test, so we got this

```
        if 1+i==1 then
            local a = (pv+fv)/n         -- = f(0) - pmt
            local b = (n*n-1)*a/12*i    -- = f''(0)/2*i
            local c = (pv-fv+a)*0.5 + b -- = f'(0) + f''(0)/2*i
            f = a + pmt + c*i
            eps = -f / (b+c)
        else ...
```

**Comment by achan001** [comment]
Regarding test 1+i==1, I have second thought.
Original 1+i*i==1 perhaps is better for general case
(rate iteration simply pass thru a tiny rate, then continue for actual rate)

Let's hold on this issue until something better.
Adding f''(0)/2*i^2 term, however, is a no-brainer.

**Comment by achan001** [comment]
Here is the problem that get into trouble, but fixed with 1+n*i^2==1 test
https://www.hpmuseum.org/forum/thread-16565-post-187993.html#pid187993

```
lua> tvm(365*24*60,nil,0,-0.01,5260.382426)
3.170979198139893e-09
lua> tvm(365*24*60,_,0,-0.01, nil)
5260.382426
```

Plus42 binary get I%YR = **3.17009**83686578233e-7
With this I, solve for FV = **5,260.38**1207980481

**Comment by achan001** [comment]
f(i) may not be accurately calculated, due to cancellation errors.
On my tvm() code, I have increased the fuzziness for no sign changes case.

Instead of (i + eps * 1e-3 == i), it now check **(i + eps * 1e-6 == i)**

I figured for true no solution case, last 2 rates would not be close at all.

> https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTq8de1dhNClmc_6dPOoTPtV815bPOD0fmD6w&usqp=CAU

**Comment by thomasokken** [comment]
Hi Albert! I would like to (finally) apply your latest fix to Plus42.
I see that the MoHPC thread you mentioned in your June 11 post has had a lot of traffic; which version of your patch would you recommend I use?

**Comment by achan001** [comment]
Test for small rate changed to (1 + n * i * i == 1)
Test for rate convergence to (i + eps * 1e-6 == i)

That's it.

**Comment by thomasokken** [comment]
I applied those two changes: https://github.com/thomasokken/plus42desktop/commit/2c7e2851f91c6e0edc879201006684f8aabff70f

I uploaded test builds for Windows to https://thomasokken.com/plus42/download/test/

**Comment by achan001** [comment]
Just tested Plus42 update on [problem 3](https://www.hpmuseum.org/forum/post-187993.html#pid187993)
Note: Dollar scaled to unit of 0.01 cent

N = 365 * 24 * 60 * 60 = 31536000
PV = 0
PMT = -100
FV = 3316670067

Plus42 Decimal, I%YR = 3.170979200111145309817810224484782e-7

For IEEE double, this rate, 1 + i * i == 1, but 1 + n * i * i ≠ 1

Old Plus42 Binary, I%YR = 3.118083226307825e-7
New Plus42 Binary, I%YR = 3.1709792001111495e-7

This reminded me, if not too much trouble.
I had made small rate calculations more accurate, fitting curve as quadratic.

```       
if 1+n*i*i==1 then
    f = (pv+fv+n*pmt)/n
    local a = f - pmt           -- = f(0) - pmt
    local b = (n*n-1)*a/6*i     -- = f''(0)*i
    fp = (pv-fv+a)*.5 + b       -- = f'(0) + f''(0)*i
    f  = f + (fp-.5*b)*i
else
```

**Comment by thomasokken** [comment]
> This reminded me, if not too much trouble, add below patch as well.
> 
> ```
> if 1+n*i*i==1 then
>     f = (pv+fv+n*pmt)/n
>     local a = f - pmt           -- = f(0) - pmt
>     local b = (n*n-1)*a/6*i     -- = f''(0)*i
>     fp = (pv-fv+a)*.5 + b       -- = f'(0) + f''(0)*i
>     f  = f + (fp-.5*b)*i
> else
> ```

Where exactly should this code be inserted, or what should it replace?

**Comment by achan001** [comment]
Patch **do_i_pct_yr()** in core_commands9.cc

Old:
```
if (1 + n * i * i == 1) {
    phloat a = (pv + fv) / n;
    phloat b = pv - fv;
    phloat fp = (a + b) / 2;
    phloat fpp = (n * n - 1) * a / 6;
    f = a + pmt + fp * i;
    fp += fpp * i;
    i += (eps = -f / fp);
} else {
```

New:
```
if (1 + n * i * i == 1) {
    f = (pv + fv + n*pmt) / n;          // f(0)
    phloat a = f - pmt;                 // f(0) - pmt
    phloat b = (n * n - 1) * a / 6 * i; // f''(0)*i    
    phloat fp = (pv - fv + a) / 2 + b;  // f'(0) + f''(0)*i
    f = f + (fp - b / 2) * i;           // f(0) + f'(0)*i + f''(0)/2*i^2
    i += (eps = -f / fp);
} else {
```

**Comment by thomasokken** [comment]
OK, thanks!
I applied the patch: https://github.com/thomasokken/plus42desktop/commit/360d81e97aef5970b6141a85d2330a430af34273
And I uploaded a Windows test build: https://thomasokken.com/plus42/download/test/

**Comment by achan001** [comment]
Tested Plus42 Binary, and it match my TVM result for borderline cases.

Thank you!

**Comment by achan001** [comment]
Sample test I did, lua tvm() vs Plus42

lua> tvm(480, nil, 2e9, -4166667, 0)
3.326403238068074e-10
lua> 480 * _ * _
5.3111800810702896e-17    -- n * i * i ≈ 1/4 (ULP of 1), close to border line
```
3.326403237606108e-10   -- tvm f = ((pv+fv)/(K-1) + pv)*i + pmt
3.326403326403326e-10   -- tvm f = f(0) + f'(0)*i

3.326403238068074e-10   -- tvm f = f(0) + f'(0)*i + f''(0)/2*i^2
3.326403238068074e-10   -- Plus42 Binary I%YR / 100
3.326403238068074375217374814559477e-10 -- Plus42 Decimal / 100
```

1 + n * i * i == 1 test is not perfect, it is only a trade-off.
Border line is where *BOTH* branches have trouble, only cure is more precisions.

**Comment by thomasokken** [comment]
Thank you!
This will be in 1.1.14, which will appear shortly.
